OpenRA.Game/Network/FrameData.cs

```diff
@@ -27,55 +27,87 @@ public override string ToString()
 			}
 		}
 
-		readonly Dictionary<int, int> clientQuitTimes = new Dictionary<int, int>();
-		readonly Dictionary<int, Dictionary<int, byte[]>> framePackets = new Dictionary<int, Dictionary<int, byte[]>>();
+		readonly HashSet<int> quitClients = new HashSet<int>();
+		readonly Dictionary<int, Queue<byte[]>> framePackets = new Dictionary<int, Queue<byte[]>>();
+		readonly Queue<int> timestepData = new Queue<int>();
 
-		public IEnumerable<int> ClientsPlayingInFrame(int frame)
+		public IEnumerable<int> ClientsPlayingInFrame()
 		{
-			return clientQuitTimes
-				.Where(x => frame <= x.Value)
-				.Select(x => x.Key)
-				.OrderBy(x => x);
+			return framePackets.Keys.Where(x => !quitClients.Contains(x)).OrderBy(x => x);
 		}
 
-		public void ClientQuit(int clientId, int lastClientFrame)
+		public void AddClient(int clientId)
 		{
-			if (lastClientFrame == -1)
-				lastClientFrame = framePackets
-					.Where(x => x.Value.ContainsKey(clientId))
-					.Select(x => x.Key).MaxByOrDefault(x => x);
+			if (!framePackets.ContainsKey(clientId))
+				framePackets.Add(clientId, new Queue<byte[]>());
+		}
+
+		public void ClientQuit(int clientId)
+		{
+			quitClients.Add(clientId);
+		}
+
+		public void AddFrameOrders(int clientId, byte[] orders, int timestep)
+		{
+			// HACK: Due to design we can actually receive client orders before the game start order
+			// has been acted on, since immediate orders are buffered, so not all clients will have
+			// been added yet. However, all clients are guaranteed to be added before the first
+			// frame is stepped since they are added in OrderManager.StartGame()
+			AddClient(clientId);
+
+			var frameData = framePackets[clientId];
+			frameData.Enqueue(orders);
+
+			if (timestep != 0)
+				timestepData.Enqueue(timestep);
+		}
+
+		public bool IsReadyForFrame()
+		{
+			return !ClientsNotReadyForFrame().Any();
+		}
+
+		public IEnumerable<int> ClientsNotReadyForFrame()
+		{
+			return ClientsPlayingInFrame()
+				.Where(client => framePackets[client].Count == 0);
+		}
 
-			clientQuitTimes[clientId] = lastClientFrame;
+		public IEnumerable<ClientOrder> OrdersForFrame(World world)
+		{
+			return ClientsPlayingInFrame()
+				.SelectMany(x => framePackets[x].Dequeue().ToOrderList(world)
+					.Select(y => new ClientOrder { Client = x, Order = y }));
 		}
 
-		public void AddFrameOrders(int clientId, int frame, byte[] orders)
+		public int BufferSizeForClient(int client)
 		{
-			var frameData = framePackets.GetOrAdd(frame);
-			frameData.Add(clientId, orders);
+			return framePackets[client].Count;
 		}
 
-		public bool IsReadyForFrame(int frame)
+		public bool TryPeekTimestep(out int timestep)
 		{
-			return !ClientsNotReadyForFrame(frame).Any();
+			if (timestepData.Any())
+			{
+				timestep = timestepData.Dequeue();
+				return true;
+			}
```

Comment:
```suggestion
			if (timestepData.TryDequeue(out timestep))
			{
				return true;
			}
			else
			{
				timestep = 0;
				return false;
			}
```

Reply:
Style nit: No braces around the first if.